### PR TITLE
X86 optimize context switch

### LINKED ISSUE #13003: context switch in x86 memory domains needs optimization
Context switching between threads in different memory domains on x86 is very slow with difficult to predict latency. Unlike other arches, which just write a fixed bank of registers, we have to go through and walk the page tables multiple times to rearrange the permissions on every page included by either the outgoing or incoming memory domain. This all happens with interrupts locked and can take varying amounts of time. This is very bad from a hard real-time perspective.

This is the code:

```
void _x86_swap_update_page_tables(struct k_thread *incoming,
				  struct k_thread *outgoing)
{
	/* Outgoing thread stack no longer accessible */
	_x86_mmu_set_flags((void *)outgoing->stack_info.start,
			   ROUND_UP(outgoing->stack_info.size, MMU_PAGE_SIZE),
			   MMU_ENTRY_SUPERVISOR, MMU_PTE_US_MASK);


	/* Userspace can now access the incoming thread's stack */
	_x86_mmu_set_flags((void *)incoming->stack_info.start,
			   ROUND_UP(incoming->stack_info.size, MMU_PAGE_SIZE),
			   MMU_ENTRY_USER, MMU_PTE_US_MASK);

	/* In case of privilege elevation, use the incoming thread's kernel
	 * stack, the top of the thread stack is the bottom of the kernel stack
	 */
	_main_tss.esp0 = incoming->stack_info.start;

	/* If either thread defines different memory domains, efficiently
	 * switch between them
	 */
	if (incoming->mem_domain_info.mem_domain !=
	   outgoing->mem_domain_info.mem_domain){

		 /* Ensure that the outgoing mem domain configuration
		  * is set back to default state.
		  */
		_arch_mem_domain_destroy(outgoing->mem_domain_info.mem_domain);
		_arch_mem_domain_configure(incoming);
	}
}
```

The x86 memory protection port is not intended for devices with constrained amounts of RAM. What we really need to do is when memory domains are initialized or modified, to allocate and configure a page table for that memory domain. Then upon context switch, instead of going through and doing expensive modification of a single page table with the two calls to _arch_mem_domain_destroy() and _arch_mem_domain_configure(), just update the CR3 register to point to the incoming thread's memory domain. This takes a fixed amount of time and is much faster.

The solution is not perfect as we still have to do some page table walking for adjusting the permission of thread stack pages for the incoming/outgoing threads, but this should still be an improvement. If we wanted to go all the way and make this completely a single step operation, we would need per-thread page tables which I feel is probably too expensive from a RAM perspective.

In more detail:

- The page table we generate at build time will be the kernel's page table, and will be used for any threads that are not part of any memory domain.

- We need to introduce a k_mem_slab pool of 4K pages, to draw from when we configure memory domains.

- When a memory domain is initialized, we will allocate some pages out of the slab pool and clone the kernel's page table into it, and save a pointer to that page table in some arch-specific field in the k_mem_pool struct (we will have to add this), since this page table is associated with the domain and not an individual thread.

- When a memory domain is configured, either at initialization or when partitions are added/removed, the domain's page table will be modified to suit the configuration. Removing partitions will just copy the old configuration for each affected page from the kernel's page table.

- On context switch:
-- We remove access to the outgoing stack like we currently do.
-- If the incoming thread is in a different domain, we load the incoming thread's domain page table into CR3, or the kernel's page table if the incoming thread is not a member of a domain
-- We grant access to the incoming thread's stack memory like we currently do.

### PR DESCRIPTION
**Pull request by tyrantlizard77**
The PR does the following:

Major changes:

- x86 now uses per-thread page tables, with a simple CR3 update on context switch or when trampolining to/from user mode
- memory domain architecture interface no longer makes decisions on whether the arch code needs to be notified, necessary when domain configuration is managed per-thread instead of common data or registers

Minor changes:

- thread->stack_obj is now set when z_new_thread() is called
- stack overflow guard pages on x86 are now read-only instead of non-present, making debugging easier
- added debug functions for dumping page tables at runtime
- added inline functions to get/set active page tables on x86
- fatal exceptions on x86 now report CR3 register (which points to the active page tables)
- memory domain documentation clarification
- some test case adjustments

Please read individual commit messages for more details.

This patch leaves the build-time page table generation alone, although this new code doesn't care how the master page tables were created and when we switch to runtime generated page tables at boot, this code can remain the same.

Addresses https://github.com/zephyrproject-rtos/zephyr/issues/15135 on x86
Fixes: https://github.com/zephyrproject-rtos/zephyr/issues/13003
Contains groundwork necessary for: #13441 #13074 #15223

### CODE REVIEW

**Comment by andrewboie** [comment]
The memory domain changes are **very** WIP, I just hacked it enough to get working. Need to study ARM/ARM some more. The memory domain changes will not only help with this PR, but also https://github.com/zephyrproject-rtos/zephyr/issues/15223 and https://github.com/zephyrproject-rtos/zephyr/issues/15135

File: arch/x86/core/ia32/x86_mmu.c [hunk 1]
```diff
@@ -225,128 +335,199 @@ void z_x86_mmu_set_flags(struct x86_mmu_pdpt *pdpt, void *ptr,
...
-{
-	struct k_mem_partition *partition;
 
-	__ASSERT_NO_MSG(domain != NULL);
-	__ASSERT(partition_id <= domain->num_partitions,
-		 "invalid partitions");
+	/* Thread isn't in user mode (yet). No need to do anything here,
+	 * this function gets called again upon actual entry into user mode
+	 */
+	if ((thread->base.user_options & K_USER) == 0) {
```
Comment (andrewboie) [inline arch/x86/core/ia32/x86_mmu.c]: simpler to get rid of this

File: arch/x86/core/ia32/x86_mmu.c [hunk 2]
```diff
@@ -225,128 +335,199 @@ void z_x86_mmu_set_flags(struct x86_mmu_pdpt *pdpt, void *ptr,
...
+		return;
+	}
 
-	partition = &domain->partitions[partition_id];
-	z_x86_reset_pages((void *)partition->start, partition->size);
-}
+	/* The page tables are at the very beginning of the stack object */
+	dst = (struct x86_mmu_pt *)thread->stack_obj;
+	if (dst == NULL) {
```
Comment (andrewboie) [inline arch/x86/core/ia32/x86_mmu.c]: check this

File: arch/x86/core/ia32/x86_mmu.c [hunk 3]
```diff
@@ -225,128 +335,199 @@ void z_x86_mmu_set_flags(struct x86_mmu_pdpt *pdpt, void *ptr,
...
+		      ROUND_UP(thread->stack_info.size, MMU_PAGE_SIZE),
+		      MMU_ENTRY_PRESENT | K_MEM_PARTITION_P_RW_U_RW,
+		      MMU_PTE_P_MASK | K_MEM_PARTITION_PERM_MASK);
 
-	partition = &domain->partitions[partition_id];
-	activate_partition(partition);
+	/* Apply memory domain configuration to these page tables, if the
+	 * thread has been assigned to one.
+	 */
+	if (mem_domain == NULL) {
```
Comment (andrewboie) [inline arch/x86/core/ia32/x86_mmu.c]: make this part a second function called on mem domain update

File: include/arch/x86/ia32/arch.h [hunk 4]
```diff
@@ -571,18 +571,62 @@ extern u32_t z_timer_cycle_get_32(void);
...
+#ifdef CONFIG_USERSPACE
+/* We need page tables for each thread in the system which runs in user mode.
+ * This is not a complete set: we just need a set of page tables which cover
+ * system RAM. When switching between threads, the relevant page directory
+ * entries will be updated to point to the incoming thread's page tables
+ * for system RAM.
+ *
+ * NOTE: We are operating under the assumption that memory domain partitions
+ * will not be configured which grant permission to address ranges outside
+ * of system RAM.
```
Comment (andrewboie) [inline include/arch/x86/ia32/arch.h]: need to update mem domain docs

File: arch/x86/Kconfig [hunk 5]
```diff
@@ -10,6 +10,7 @@ menu "X86 Architecture Options"
 
 config ARCH
 	default "x86"
+	select NEW_MEM_DOMAIN_ARCH_INTERFACE
```
Comment (andrewboie) [inline arch/x86/Kconfig]: move to arch/Kconfig

File: arch/x86/core/ia32/thread.c [hunk 6]
```diff
@@ -42,6 +42,160 @@ struct _x86_initial_frame {
...
+	}
+}
+
+/* Track the last user thread we swapped in */
+static struct k_thread *last_user;
+
+void _x86_swap_update_page_tables(struct k_thread *incoming,
+				  struct k_thread *outgoing)
+{
+	if (incoming == outgoing) {
```
Comment (andrewboie) [inline arch/x86/core/ia32/thread.c]: Really seems like a swap bug

**Comment by andrewboie** [comment]
Still WIP, but correctly AFAICT implements cr3 updates on context switch for switching between page tables.

TODO
- Still need to properly amend memory domain APIs, what we have here is hacked up
- Issues with dynamically created thread objects that I'm not sure the best way to resolve
- Issues with greater stack usage by k_thread_create() due to page table construction, had to hack main thread stack size on the stack canary test
- Documentation updates

@andyross I think this is ready for you to have a look at, the core page table stuff should be correct.

After this goes in I'm going to work on creating z_x86_kernel_pdpt/z_x86_user_pdpt dynamically at boot instead of build time like we discussed

File: arch/x86/core/ia32/swap.S [hunk 7]
```diff
@@ -153,15 +153,13 @@ SECTION_FUNC(TEXT, __swap)
 #ifdef CONFIG_X86_IAMCU
 	push	%eax
 #else
-	push	%edx
 	push	%eax
 #endif
-	call	_x86_swap_update_page_tables
+	call	z_x86_swap_update_page_tables
```
Comment (andrewboie) [inline arch/x86/core/ia32/swap.S]: better to not call this at all if KPTI since it does nothing

File: arch/x86/include/ia32/mmustructs.h [hunk 8]
```diff
@@ -526,6 +526,19 @@ struct x86_mmu_pt {
  */
 void z_x86_dump_page_tables(struct x86_mmu_pdpt *pdpt);
 
+static inline void z_x86_page_tables_set(struct x86_mmu_pdpt *pdpt)
```
Comment (andyross) [inline arch/x86/include/ia32/mmustructs.h]: Style: I'm not sure an abstracted name is actually helping here.  Setting CR3 is a big event (especially with virtual memory, but even when just changing permissions) and personally I'd prefer to see this called out really clearly and loudly in code rather than being a mild-seeming "set page tables" call.

I mean... doing this from C at all seems like it's likely to be a maintenance danger long term: it really only makes sense with our existing unified physical memory space.  Changing to a VM won't ever make sense with a call like this.   Maybe it should be part of the saved context frame that gets restored by dedicated assembly right before the IRET?  Where else are we calling it?

File: kernel/Kconfig [hunk 9]
```diff
@@ -499,6 +499,12 @@ config HEAP_MEM_POOL_MIN_SIZE
 	  of the entire pool.
 endmenu
 
+config ARCH_HAS_ALIGNED_THREAD_OBJECTS
```
Comment (andyross) [inline kernel/Kconfig]: Not loving this.  This is a quirk specific to x86 fxsave/fxrstor, and IMHO not really something generic.  If we want to disallow this feature on x86, let's just do "depends on !X86" and document the limitation somewhere, so that later on we can fix it (by using normal I/O instructions, or doing the fxsave to a location that isn't part of the thread struct like an aligned region of the kernel stack, or a separate region, etc...).  Or just make the _thraed.arch field a pointer and have the arch define its own region with its own requirements and initialize it, etc...

Basically I have a hard time imagining this ever used for anything but this current workaround, and would prefer the Zephyr thread struct not have this kind of arch-specific thing in it.

File: include/arch/x86/ia32/arch.h [hunk 10]
```diff
@@ -571,18 +571,84 @@ extern u32_t z_timer_cycle_get_32(void);
...
+ * macros to compute how many pages we need.
+ */
+
+/* Define a range [Z_X86_PT_START, Z_X86_PT_END) which is the memory range
+ * covered by all the page tables needed for system RAM
+ */
+#define Z_X86_PT_START	((u32_t)ROUND_DOWN(DT_PHYS_RAM_ADDR, Z_X86_PT_AREA))
+#define Z_X86_PT_END	((u32_t)ROUND_UP(DT_PHYS_RAM_ADDR + \
+					 (DT_RAM_SIZE * 1024U), \
+					 Z_X86_PT_AREA))
```
Comment (andyross) [inline include/arch/x86/ia32/arch.h]: If the start address is rounded down, then it can be lower than DT_PHYS_RAM_ADDR, likewise the "END" address may be higher than the system RAM limit.  Is that not an overflow?  What am I missing here?

I think the idea here is that you want to detect the maximum bounding set of page addresses that encompass system RAM so that all addresses "can" be accessed through the page tables, but if (for some crazy reason) it's not actually page aligned naturally, and we actually try to use those pages for anything, then we'll end up out of bounds and anything can happen.

Obviously in practice all systems have page-aligned memory regions because that's what the hardware supports, and if we ever hit the case where it's not it's surely going to be because of a typo in the devicetree files.

This doesn't really sound like a useful feature.  I'd suggest turning this into a build assert that guarantees that the devicetree memory is page-aligned, or if some some reason we have to handle that on some hardware, we should be swapping UP/DOWN above so that the PT limits are always guaranteed to be in bounds.

**Review by andyross — COMMENTED** [review]
Some notes, mostly nitpicking.  Still need to dig through the allocation code, but can't find anything actually incorrect.

File: kernel/Kconfig (hunk 9 above)
Comment (andrewboie) [reply kernel/Kconfig]: > Not loving this

Yeah I don't like it either.

> Or just make the _thraed.arch field a pointer and have the arch define its own region with its own requirements and initialize it, etc...

Pointer to what memory? I don't want to bring any kind of heap allocations into the picture.

> This is a quirk specific to x86 fxsave/fxrstor

Also applies to the user_pdpt which I put in the thread->arch struct, which is why I did this.

The problem is, if I can't put the pdpt in thread->arch, where can I put it?

- Can't go in the thread stack, user mode must never be able to modify it
- Can't go in the kernel stack, as if the kernel stack fills up to the point where it is just short of the guard page, the user_pdpt could get silently corrupted
- Could allocate yet another page for the pdpt in the stack object, but that seems a dreadful waste of RAM

Since we are not RAM constrained on our currently supported x86 targets, I think I'll just add a dedicated page to the stack object for the PDPT and file an enhancement to optimize it later. One idea I had is to put it in the area for the stack overflow guard, and adjust the MMU permission on that page when pdpt is modified (happens only in one spot). However USERSPACE and !HW_STACK_PROTECTION is a valid configuration. I guess I could allocate that page unconditionally.

Comment (andyross) [reply kernel/Kconfig]: Well... the use case is dynamic object, right?  So dynamic allocation for the benefit of those seems acceptable.  It could even be simple, like have a mem_slab of aligned data which were allocated first-come-first-serve for new threads.

For static ones, you could have a Z_ARCH_THREAD_DATA_DEFINE or whatever macro defined by the arch, and put the arch struct in an arch-controlled section with arch-controlled alignment, and then initialize a pointer field in the thread struct.  Has the cost of the pointer I guess, though even that could be optimized out with a z_arch_thread_data() call or whatever that just returns an offset from the thread pointer on architectures where you can put it right after the thread struct.

My whine here was really just about having an official-looking "HAS_ALIGNED_THREAD_OBJECTS" kconfig, when really this seems like a treatable design problem we could special case by disabling for now and filing a github issue with "x86 doesn't support dynamic thread objects" to fix later.

File: arch/x86/include/ia32/mmustructs.h (hunk 8 above)
Comment (andrewboie) [reply arch/x86/include/ia32/mmustructs.h]: > Style: I'm not sure an abstracted name is actually helping here. Setting CR3 is a big event (especially with virtual memory, but even when just changing permissions) and personally I'd prefer to see this called out really clearly and loudly in code rather than being a mild-seeming "set page tables" call.

I don't agree with you at all. I think instead of putting snippets of assembly code directly in C functions, they should be abstracted to inline functions that do the asm magic. 

MISRA-C agrees with me. See Directive 4.3 Assembly Language shall be encapsulated and isolated.

> I mean... doing this from C at all seems like it's likely to be a maintenance danger long term: it really only makes sense with our existing unified physical memory space. Changing to a VM won't ever make sense with a call like this. 

There's no maintenance danger, this stuff either passes our tests or it doesn't.

> Where else are we calling it?

A CR3 update needs to happen in either two places:

1. If KPTI is disabled, then we update CR3 on context switch. This is where this inline function gets called from z_x86_swap_update_page_tables(), which is invoked from __swap().
2. If KPTI is enabled, then we don't need to update CR3 on context switch; supervisor threads, or threads elevated to privileged mode, all use the global kernel pdpt. Instead, when swap is done and if we trampoline back to user mode via z_x86_trampoline_to_user, CR3 gets set from the kernel's page tables to the per-thread user page tables right before iret.

File: kernel/Kconfig (hunk 9 above)
Comment (andrewboie) [reply kernel/Kconfig]: > Can't go in the kernel stack, as if the kernel stack fills up to the point where it is just short of the guard page, the user_pdpt could get silently corrupted

Wait, actually it can.
Store it at the HIGH addresses of the kernel stack, and on elevation the initial kernel stack pointer just needs to be below it.

I'm on it.

File: include/arch/x86/ia32/arch.h (hunk 10 above)
Comment (andrewboie) [reply include/arch/x86/ia32/arch.h]: This isn't for page alignment. All addresses are assumed to be page aligned.
This is the bounds of memory covered by a *page table*, which is 2 megabytes.
This ensures that (for whatever reason) if say RAM starts at 1MB, and we have 4 MB of RAM, then we generate three page user page tables covering 0-2MB, 2MB-4MB, 4MB-6MB, since RAM spans from 1MB-5MB.

File: include/arch/x86/ia32/arch.h [hunk 11]
```diff
@@ -571,18 +571,84 @@ extern u32_t z_timer_cycle_get_32(void);
...
- * +---------------+
+ * +---------------------------------+
+ * | Thread stack                    |
+ * +---------------------------------+
+ * | Kernel stack                    |
+ * +---------------------------------+
+ * | Guard page                      |
+ * +---------------------------------+
+ * | User page tables                |
+ * +---------------------------------+
```
[the remark this answers is not in the document]
Comment (andrewboie) [reply include/arch/x86/ia32/arch.h]: That's a great idea! It will make the code much clearer. I'll file a story for 2.1.

File: arch/x86/include/ia32/mmustructs.h (hunk 8 above)
Comment (andyross) [reply arch/x86/include/ia32/mmustructs.h]: To be clear: I wasn't asking for inline assembly, just for this to be called out in a really obvious way.  It's not a C function, it can't be implemented in C, and in the general case of modified mappings can't even be survived when called from C.

I just don't think "set page tables" properly captures what is happening to a reader.  It didn't to me.

Comment (andrewboie) [reply arch/x86/include/ia32/mmustructs.h]: I'm not sure then what you want here.

Is your desire that z_x86_swap_update_page_tables() be reimplemented in assembly and place inline with __swap()?

I try, whenever possible, to implement things in C rather than ASM, the compiler is much smarter than I in creating tight asm code.

Comment (andyross) [reply arch/x86/include/ia32/mmustructs.h]: I think if this was called out in code with a big comment to the effect of "This changes CR3, no memory access below is legal unless it is known for sure that the mappings are identical!" (and maybe a similar comment in __swap once it returns to make sure it's touching only the stack) and had a name like "z_x86_set_cr3()", then I'd have nothing to complain about.

Right now there's a comment that mentions CR3, but only in the context of a performance optimization, and I saw that and went "wait, what? This changes CR3?".  Thus the comment here.

Stated differently: right now, _x86_swap_update_page_tables() is piecewise changing mappings that we can reasonably "know" are not related to the current code, where with CR3 you change them all at once.  And as time goes on and we support more stuff like virtual memory, it's going to be changing things into weirder and weirder environments.

Comment (andrewboie) [reply arch/x86/include/ia32/mmustructs.h]: > I think if this was called out in code with a big comment to the effect of "This changes CR3, no memory access below is legal unless it is known for sure that the mappings are identical!" (and maybe a similar comment in __swap once it returns to make sure it's touching only the stack) and had a name like "z_x86_set_cr3()", then I'd have nothing to complain about.

This is all agreeable to me, will do.

File: arch/x86/include/ia32/mmustructs.h [hunk 12]
```diff
@@ -491,13 +499,46 @@ union x86_mmu_pde {
...
+ *
+ * Entry codes in uppercase indicate that user mode may access.
+ *
+ * @param pdpt Top-level pointer to the page tables, as programmed in CR3
+ */
+void z_x86_dump_page_tables(struct x86_mmu_pdpt *pdpt);
+
+static inline void z_x86_page_tables_set(struct x86_mmu_pdpt *pdpt)
+{
+	__asm__ volatile("movl %0, %%cr3\n\t" : : "r" (pdpt));
```
Comment (gnuless) [inline arch/x86/include/ia32/mmustructs.h]: Do you need a memory barrier here?  In case the compiler decides to get fancy and re-order a store around the page-table switch.

Comment (andrewboie) [reply arch/x86/include/ia32/mmustructs.h]: Good question. I am unsure, but I'll be safe and add one.

**Comment by andrewboie** [comment]
Update:

- PDPT now stored in the stack object, sidestepping any issues with dynamic thread. The patch which disables dynamic threads on X86 has been dropped.

- Documentation updates

- Moved inline function to set CR3 out of the header and into x86_mmu.c. Memory barrier added.

- Additional comments

File: arch/x86/include/ia32/mmustructs.h (hunk 12 above)
Comment (andrewboie) [reply arch/x86/include/ia32/mmustructs.h]: Done. This function is now in thread.c, with a barrier and additional comments added.

File: arch/x86/include/ia32/mmustructs.h (hunk 8 above)
Comment (andrewboie) [reply arch/x86/include/ia32/mmustructs.h]: Done. This function is now in thread.c, with a barrier and additional comments added.

File: kernel/Kconfig (hunk 9 above)
Comment (andrewboie) [reply kernel/Kconfig]: PDPT moved to the stack object, ARCH_HAS_ALIGNED_THREAD_OBJECTS patch dropped.

**Comment by andrewboie** [comment]
Update:

- Specify x86 stack object layout using a struct based on @gnuless suggestion

**Review by ghost — APPROVED** [review]

**Review by dbkinder — COMMENTED** [review]
No doc changes to review

**Comment by andrewboie** [comment]
@ioannisg @vonhust @ruuddw any objections to the two patches which touch the mem domain API and documentation changes? they are fairly trivial, mostly just moved the current thread check into the arch implementations

**Review by wentongwu — APPROVED** [review]

**Comment by ioannisg** [comment]
> 
> 
> @ioannisg @vonhust @ruuddw any objections to the two patches which touch the mem domain API and documentation changes? they are fairly trivial, mostly just moved the current thread check into the arch implementations
> 
> The relevant patches in this PR are
> 
>     * "userspace: adjust arch memory domain interface"
> 
>     * "userspace: clarify k_mem_partition_add()"

@andrewboie sorry for being late on this.
I had a look in the mem_domain related changes, and they reflect your intention of moving the decisions to arch-specific code. So I think they are fine.

So, AFAICS, it is the arch-specific code which decides to re-program MPU (ARC, ARM), if the functions are called for the current thread. Does this mean that this principle is no longer needed, generally, e.g. for other architectures, such as x86?

In that case, I believe we lack some documentation in arm_core_mpu.c, arc_core_mpu.c, or in the corresponding API header, that clarifies why these checks are required. IMHO, maybe some notes on the function docs would suffice...

**Comment by andrewboie** [comment]
> So, AFAICS, it is the arch-specific code which decides to re-program MPU (ARC, ARM), if the functions are called for the current thread. 

Not the current thread. Now it's any thread.

>

File: include/kernel.h [hunk 13]
```diff
@@ -4938,7 +4941,22 @@ extern void k_mem_domain_destroy(struct k_mem_domain *domain);
...
+ *
+ * - Partition bounds must be within system RAM boundaries on MMU-based
+ *   systems.
+ * - Partitions in the same memory domain may not overlap each other.
+ * - Partitions must not be defined which expose private kernel
+ *   data structures or kernel objects.
+ * - The starting address alignment, and the partition size must conform to
+ *   the constraints of the underlying memory management hardware, which
+ *   varies per architecture.
+ * - Memory domain partitions are only intended to control access to mamory
```
Comment (ruuddw) [inline include/kernel.h]: mamory -> memory
few lines below: Volating -> Violating

**Comment by ruuddw** [comment]
Moving the check for 'MPU HW' reprogramming or not to the arch specific code is ok for me if it helps other arches. Not sure yet if it would help or be required for  #13074 - managing such a 'TLB' would preferably be generic and not arch specific code. I can imagine a 3 layer approach instead of current 2 layers as well: logical/api mem domain management -> logical to physical mapping (and lazy programming/TLB management) -> physical MPU programming (as in current arch specific code).

**Comment by andrewboie** [comment]
> Moving the check for 'MPU HW' reprogramming or not to the arch specific code is ok for me if it helps other arches. Not sure yet if it would help or be required for #13074 - managing such a 'TLB' would preferably be generic and not arch specific code.

That doesn't matter, it manages per-thread data so it will require this

**Comment by ioannisg** [comment]
>I'm not quite sure what you mean, it depends on the arch code. On x86 it's any thread because there is arch-specific state that has to be managed for any thread that was in the domain being modified (in this case the per-thread pagetables)

So if I destroy a memory domain, I need to update all the per-thread page tables for the threads that were running in it, because they are no longer in a memory domain.

OK, so I understand that ARC and ARM only call the on-line reprogramming if it is the current thread, while in x86 it is any thread. Thanks for clarifying.

So, as I said above (maybe not clear enough) is that I would like _us_ to document the reasoning why we only need to do on-line re-programming if it is the current thread in ARM Core MPU code. Earlier, these checks for current-thread were not present in arch-code, because they were in kernel/mem_domain, but since they are now in ARM core mpu, i think it would be nice to comment why we do them, to help _explicitly_ the ARM developer understand the code better.

**Review by ioannisg — APPROVED** [review]

**Review by ruuddw — APPROVED** [review]

**Review by carlescufi — APPROVED** [review]
